// ternoa-js/helpers/encryption.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.decryptFile = exports.encryptFile = exports.encryptContent = exports.generatePGPKeys = void 0;
var openpgp = __importStar(require("openpgp"));
var utils_1 = require("./utils");
/**
 * @name generatePGPKeys
 * @summary                 Generates a new PGP key pair.
 * @returns                 An object with both private and public PGP keys.
 */
var generatePGPKeys = function () { return __awaiter(void 0, void 0, void 0, function () {
    var _a, privateKey, publicKey;
    return __generator(this, function (_b) {
        switch (_b.label) {
            case 0: return [4 /*yield*/, openpgp.generateKey({
                    type: "ecc",
                    curve: "curve25519",
                    userIDs: [{ name: "Jon Smith", email: "jon@example.com" }],
                })];
            case 1:
                _a = _b.sent(), privateKey = _a.privateKey, publicKey = _a.publicKey;
                return [2 /*return*/, { privateKey: privateKey, publicKey: publicKey }];
        }
    });
}); };
exports.generatePGPKeys = generatePGPKeys;
/**
 * @name encryptContent
 * @summary                 Encrypts a content (string).
 * @param content           Content to encrypt.
 * @param publicPGPKey      Public Key to encrypt the content.
 * @see                     Learn more about encryption {@link https://docs.openpgpjs.org/global.html#encrypt here}.
 * @returns                 A string containing the encrypted content.
 */
var encryptContent = function (content, publicPGPKey) { return __awaiter(void 0, void 0, void 0, function () {
    var message, publicKey, encryptedContent;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, openpgp.createMessage({
                    text: content,
                })];
            case 1:
                message = _a.sent();
                return [4 /*yield*/, openpgp.readKey({
                        armoredKey: publicPGPKey,
                    })];
            case 2:
                publicKey = _a.sent();
                return [4 /*yield*/, openpgp.encrypt({
                        message: message,
                        encryptionKeys: [publicKey],
                    })];
            case 3:
                encryptedContent = _a.sent();
                return [2 /*return*/, encryptedContent];
        }
    });
}); };
exports.encryptContent = encryptContent;
/**
 * @name encryptFile
 * @summary                 Encrypts file with the public key.
 * @param file              File to encrypt.
 * @param publicPGPKey      Public Key to encrypt the file.
 * @see                     Learn more about encryption {@link https://docs.openpgpjs.org/global.html#encrypt here}.
 * @returns                 A string containing the encrypted file.
 */
var encryptFile = function (file, publicPGPKey) { return __awaiter(void 0, void 0, void 0, function () {
    var buffer, content, encryptedFile;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, (0, utils_1.convertFileToBuffer)(file)];
            case 1:
                buffer = _a.sent();
                content = buffer.toString("base64");
                return [4 /*yield*/, (0, exports.encryptContent)(content, publicPGPKey)];
            case 2:
                encryptedFile = _a.sent();
                return [2 /*return*/, encryptedFile];
        }
    });
}); };
exports.encryptFile = encryptFile;
/**
 * @name decryptFile
 * @summary                 Decrypts message with the private key.
 * @param encryptedMessage  Message to decrypt.
 * @param privatePGPKey     Private Key to decrypt the message.
 * @see                     Learn more about encryption {@link https://docs.openpgpjs.org/global.html#decrypt here}.
 * @returns                 A base64 string containing the decrypted message.
 */
var decryptFile = function (encryptedMessage, privatePGPKey) { return __awaiter(void 0, void 0, void 0, function () {
    var privateKey, message, decryptedMessage;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, openpgp.readPrivateKey({ armoredKey: privatePGPKey })];
            case 1:
                privateKey = _a.sent();
                return [4 /*yield*/, openpgp.readMessage({
                        armoredMessage: encryptedMessage,
                    })];
            case 2:
                message = _a.sent();
                return [4 /*yield*/, openpgp.decrypt({
                        message: message,
                        decryptionKeys: privateKey,
                    })];
            case 3:
                decryptedMessage = (_a.sent()).data;
                return [2 /*return*/, decryptedMessage];
        }
    });
}); };
exports.decryptFile = decryptFile;
